engine/runtime/src/main/java/org/enso/interpreter/node/callable/InvokeMethodNode.java

```diff
@@ -154,29 +174,81 @@ private boolean typeCanOverride(MethodRootNode node, EnsoContext ctx) {
     Type warning = builtins.warning();
     Type panic = builtins.panic();
     return methodOwnerType.isEigenType()
-            && builtins.nothing() != methodOwnerType
-            && any.getEigentype() != methodOwnerType
-            && panic.getEigentype() != methodOwnerType
-            && warning.getEigentype() != methodOwnerType;
+        && builtins.nothing() != methodOwnerType
+        && any.getEigentype() != methodOwnerType
+        && panic.getEigentype() != methodOwnerType
+        && warning.getEigentype() != methodOwnerType;
   }
 
   @Specialization(
-          replaces = "doFunctionalDispatchCachedSymbol",
-          guards = {"typesLibrary.hasType(self)", "!typesLibrary.hasSpecialDispatch(self)"})
+      replaces = "doFunctionalDispatchCachedSymbol",
+      guards = {"typesLibrary.hasType(self)", "!typesLibrary.hasSpecialDispatch(self)"})
   Object doFunctionalDispatchUncachedSymbol(
-          VirtualFrame frame,
-          State state,
-          UnresolvedSymbol symbol,
-          Object self,
-          Object[] arguments,
-          @CachedLibrary(limit = "10") TypesLibrary typesLibrary,
-          @Cached MethodResolverNode methodResolverNode) {
+      VirtualFrame frame,
+      State state,
+      UnresolvedSymbol symbol,
+      Object self,
+      Object[] arguments,
+      @CachedLibrary(limit = "10") TypesLibrary typesLibrary,
+      @Cached MethodResolverNode methodResolverNode) {
     Type selfTpe = typesLibrary.getType(self);
     Function function = resolveFunction(symbol, selfTpe, methodResolverNode);
     if (function == null) {
       throw new PanicException(
-              EnsoContext.get(this).getBuiltins().error().makeNoSuchMethod(self, symbol), this);
+          EnsoContext.get(this).getBuiltins().error().makeNoSuchMethod(self, symbol), this);
+    }
+    var resolvedFuncArgCount = function.getSchema().getArgumentsCount();
+    CallArgumentInfo[] invokeFuncSchema = invokeFunctionNode.getSchema();
+    // Check if this is a static method call on Any
+    if (isAnyEigenType(selfTpe) && arguments.length == resolvedFuncArgCount - 1) {
+      // function is a static method on Any, so the first two arguments in `invokeFuncSchema`
+      // represent self arguments.
+      boolean selfArgSpecified = false;
+      if (invokeFuncSchema.length > 1) {
+        selfArgSpecified =
+            invokeFuncSchema[1].getName() != null && invokeFuncSchema[1].getName().equals("self");
+      }
+
+      if (selfArgSpecified) {
+        // If there is a self named argument in the method call, we fall back to the old
+        // behavior - there will be no prepended Any.type self argument
+        assert arguments.length == invokeFunctionNode.getSchema().length;
+        return invokeFunctionNode.execute(function, frame, state, arguments);
+      }
+
+      Object[] argsWithPrependedSelf = new Object[arguments.length + 1];
+      argsWithPrependedSelf[0] = EnsoContext.get(this).getBuiltins().any().getEigentype();
```

Comment:
OKeyish if it works, but: Ideally the `self` should be set to the actual type which we are calling the function upon. When one does `Boolean.to_text` the `self` should be `Boolean`. I am not exactly sure what would be the consequences, but such behavior would fit the model I have in my had more and it could potentially simplify some code introduced by @hubertp to allow calling `Boolean.to_text`.

Reply:
`argsWithPrependedSelf` is `Object[]` that is passed to a `InvokeFunctionNode` that represents a static method call. These nodes expect *two self arguments* - the first one is always the eigen type of the type where the method is defined - in this case `Any.type`, the second `self` can be set by the user.

> When one does Boolean.to_text the self should be Boolean

And it is. This is the second `self` argument. In this line, I am setting the first `self` argument, which should be Any eigen type.